sql: parallelize cross-range point scans

jordanlewis opened the pull request:
Previously, scans that contained multiple point spans (spans that are
guaranteed to return at most one row) were always done serially - the
spans would be chunked up by range, and each range's ScanRequests would
be sent out one by one.

This was inefficient, especially when the gateway node wasn't nearby the
node with the data required.

Now, in the case where all of the spans are guaranteed to return at most
one row and there's no limit on the scan, we permit the DistSender to
use parallelism to send out the ScanRequests.

Fixes #30943.

Release note (performance improvement): cross-range disjunctive point
scans are now automatically parallelized.

RaduBerinde committed:
opt: plumb "max results" guarantee

This guarantee is calculated in buildScan through indexConstraintMaxResults.
We can calculate the maximum number of results when both of the
following are satisfied:
1. The index columns form a strict key or a lax key and the spans do not
   specify NULL values.
2. All spans cover all the columns of the index and have
   equal start and end keys up to but not necessarily including the
   last column.

Note the "not including the last column" in the second point. If the
spans do not have equal keys on the last column, a difference is
calculated only for int types.

Release note: None

jordanlewis committed:
sql: parallelize cross-range scans when maxResults is specified

Previously, scans that contained multiple spans were always done
serially - the spans would be chunked up by range, and each
range's ScanRequests would be sent out one by one.

This was inefficient, especially when the gateway node wasn't nearby the
node with the data required.

Now, in the case where all of the spans are guaranteed to return
at most maxResults and there's no limit on the scan, we permit
the DistSender to use parallelism to send out the ScanRequests.

Release note (performance improvement): cross-range disjunctive scans
where the result size can be deduced are now automatically
parallelized.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/31616)
<!-- Reviewable:end -->

jordanlewis commented:
Couple things missing here still -

1. the heuristic planner needs to plumb through the point spans information, since it's still used to plan mutation queries.
2. need to add a trace test or other such thing that asserts that the batchrequest that actually gets emitted in the end is parallelizable.

jordanlewis commented:
Alright, this should be good to go.

One weird thing for @RaduBerinde to take a look at is that the opt "point spans" detector seems to fail in the case of unique indexes - I added a logic test for that but commented it out. It works in the equivalent code I added for the heuristic planner (had to do that because mutations still get planned that way).

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LP9FL4qDRnVTVic-2AI:-LP9FL4qDRnVTVic-2AJ:b-xrii5p) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07bda5e207c19591517fcba0d5e5d1daffbbee6/pkg/sql/opt/exec/execbuilder/relational_builder.go#L287)):*
> ```Go
> 	}
> 
> 	if !scan.Relational().FuncDeps.ColsAreStrictKey(indexCols) {
> ```

I think this is the problem. There are two ways for indexCols to form a strict key:
 - the cols form a strict key in the FuncDeps; or
 - the cols form a weak key (`FuncDeps.ColsAreLaxKey()`) and they are all non-null (`Relational().NotNullCols.Contains(indexCols)`).

We're missing the second check - try adding that and see if it fixes that test.
___
*[pkg/sql/opt/exec/execbuilder/testdata/aggregate, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LP9EnaE0NCSftOffI3b:-LP9EnaE0NCSftOffI3c:buvkgl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07bda5e207c19591517fcba0d5e5d1daffbbee6/pkg/sql/opt/exec/execbuilder/testdata/aggregate#L342)):*
> ```
> ·          table        xyz@xy             ·          ·
> ·          spans        /0-/1 /4-/5 /7-/8  ·          ·
> ·          parallel     false              ·          ·
> ```

Instead of having this show up everywhere, I would just show it only when it's true (as `parallel`).


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LP9FL4qDRnVTVic-2AI:-LPBQ8YM42cjXwlLAdeZ:b-7fvbnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07bda5e207c19591517fcba0d5e5d1daffbbee6/pkg/sql/opt/exec/execbuilder/relational_builder.go#L287)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this is the problem. There are two ways for indexCols to form a strict key:
 - the cols form a strict key in the FuncDeps; or
 - the cols form a weak key (`FuncDeps.ColsAreLaxKey()`) and they are all non-null (`Relational().NotNullCols.Contains(indexCols)`).

We're missing the second check - try adding that and see if it fixes that test.
</blockquote></details>

Actually, scratch that, there's something else wrong here (it looks like a limitation of the func deps). We're trying to figure it out.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LP9FL4qDRnVTVic-2AI:-LPBVvVN1oMtORcLQ4jS:bviwo14) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07bda5e207c19591517fcba0d5e5d1daffbbee6/pkg/sql/opt/exec/execbuilder/relational_builder.go#L287)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Actually, scratch that, there's something else wrong here (it looks like a limitation of the func deps). We're trying to figure it out.
</blockquote></details>

I filed https://github.com/cockroachdb/cockroach/issues/31628.  Please refer to that in the failing test. I will clean up when that's fixed.

Do you plan to backport this to 2.1.x?


<!-- Sent from Reviewable.io -->

nvb reviewed (changes requested):
:man_scientist: :dog: on most of this, but really happy to see it done so quickly. Thanks!

> Reviewed 1 of 3 files at r1, 23 of 23 files at r3, 16 of 16 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LPByml15bps5E8jchUv:buptftk)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBwwwP1qmUbscRifrM:-LPBwwwQ4JfAGxKK3j8q:b-xe0h88) ([raw file](https://github.com/cockroachdb/cockroach/blob/df4c452d652d271a0bc927cd184802c50bca7b23/pkg/sql/opt_exec_factory.go#L76)):*
> ```Go
> ) (exec.Node, error) {
> 
> 	// TODO(jordan): do something with this.
> ```

I wouldn't leave this in the commit.
___
*[pkg/sql/opt_index_selection.go, line 295 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBxhLG-QQeGeEGax48:-LPBxhLG-QQeGeEGax49:b-p2lkg6) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/opt_index_selection.go#L295)):*
> ```Go
> // allPointSpans returns true if the constraint only contains spans that can
> // return at most one row apiece.
> func allPointSpans(
> ```

Might be worth pointing this at the optimizer equivalent version.
___
*[pkg/sql/opt_index_selection.go, line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBxrr92SRuGzHUVZau:-LPBxrr92SRuGzHUVZav:b-s71foe) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/opt_index_selection.go#L305)):*
> ```Go
> 	}
> 	// We have point spans when both of the following are satisfied:
> 	//  1. The index is unique.
> ```

Why does the index need to be unique for this to be true?
___
*[pkg/sql/scan.go, line 43 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPByEib3C96Cmy4faFP:-LPByEib3C96Cmy4faFQ:b4ix8t0) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/scan.go#L43)):*
> ```Go
> // A scanNode handles scanning over the key/value pairs for a table and
> // reconstructing them into rows.
> type scanNode struct {
> ```

You mentioned yesterday that `scanNode` is no longer used. Is that not the case?
___
*[pkg/sql/scan.go, line 264 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBy2PxAd6nuQ4HtR-w:-LPBy2PxAd6nuQ4HtR-x:bw8wolz) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/scan.go#L264)):*
> ```Go
> // distSender level safely.
> func (n *scanNode) canParallelize() bool {
> 	return n.allPointSpans && n.hardLimit == 0 && n.softLimit == 0
> ```

use `n.limitHint() == 0` here to avoid duplicating logic
___
*[pkg/sql/logictest/testdata/planner_test/update, line 249 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPByjOo8w5GiFVyeAYM:-LPByjOo8w5GiFVyeAYN:bfd2fml) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/logictest/testdata/planner_test/update#L249)):*
> ```
> rows affected: 1
> 
> # Ensure that the scan for cross-range point updates are parallelized.
> ```

Could you add a test for DELETEs as well?
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 291 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBxIOdELg-SsA2LPVO:-LPBxIOdELg-SsA2LPVP:brkmls1) ([raw file](https://github.com/cockroachdb/cockroach/blob/df4c452d652d271a0bc927cd184802c50bca7b23/pkg/sql/opt/exec/execbuilder/relational_builder.go#L291)):*
> ```Go
> 	// the cols form a weak key and they are all non-null.
> 	//
> 	// IF neither of these conditions hold, we don't have all point spans.
> ```

s/IF/If/
___
*[pkg/sql/opt/exec/execbuilder/testdata/aggregate, line 342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LP9EnaE0NCSftOffI3b:-LPBxWl_9y2IBnQ-8n-W:b-cy0fm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07bda5e207c19591517fcba0d5e5d1daffbbee6/pkg/sql/opt/exec/execbuilder/testdata/aggregate#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Instead of having this show up everywhere, I would just show it only when it's true (as `parallel`).
</blockquote></details>

Yeah, I think that's a good idea.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained

___
*[pkg/sql/opt_exec_factory.go, line 76 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBwwwP1qmUbscRifrM:-LPCuXJp4359RkHRqZsb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df4c452d652d271a0bc927cd184802c50bca7b23/pkg/sql/opt_exec_factory.go#L76)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I wouldn't leave this in the commit.
</blockquote></details>

Done.
___
*[pkg/sql/opt_index_selection.go, line 295 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBxhLG-QQeGeEGax48:-LPCuWUx4BSEb0pNXHMp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/opt_index_selection.go#L295)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Might be worth pointing this at the optimizer equivalent version.
</blockquote></details>

Done.
___
*[pkg/sql/opt_index_selection.go, line 305 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBxrr92SRuGzHUVZau:-LPCuZl87dtjGi2R4WD5:b-a58o9x) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/opt_index_selection.go#L305)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Why does the index need to be unique for this to be true?
</blockquote></details>

Because otherwise the constraint isn't strong enough to ensure that at most one row will return from the scan.
___
*[pkg/sql/scan.go, line 43 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPByEib3C96Cmy4faFP:-LPCuhub06eU4qJ0yRaf:b-86d0ej) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/scan.go#L43)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

You mentioned yesterday that `scanNode` is no longer used. Is that not the case?
</blockquote></details>

It's used in the planning phase still, but not in the execution phase. We'll be deleting this code over time, and it will become easier as the optimizer team subsumes physical planning and plans directly from the memo to distsql structures.
___
*[pkg/sql/scan.go, line 264 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBy2PxAd6nuQ4HtR-w:-LPCuh-j1EyqMtmiS-OU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/scan.go#L264)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

use `n.limitHint() == 0` here to avoid duplicating logic
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/planner_test/update, line 249 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPByjOo8w5GiFVyeAYM:-LPCv83g1v-R4CzEs9QN:bcvjgzp) ([raw file](https://github.com/cockroachdb/cockroach/blob/87f63daac3d5ac6fc662b3a629030b2fce11d4cf/pkg/sql/logictest/testdata/planner_test/update#L249)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Could you add a test for DELETEs as well?
</blockquote></details>

Good idea. Done.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 287 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LP9FL4qDRnVTVic-2AI:-LPCvJy1B8dJbHxcww4g:b-cm0idp) ([raw file](https://github.com/cockroachdb/cockroach/blob/e07bda5e207c19591517fcba0d5e5d1daffbbee6/pkg/sql/opt/exec/execbuilder/relational_builder.go#L287)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I filed https://github.com/cockroachdb/cockroach/issues/31628.  Please refer to that in the failing test. I will clean up when that's fixed.

Do you plan to backport this to 2.1.x?
</blockquote></details>

Done. Yes, this will need to be backported.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 291 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPBxIOdELg-SsA2LPVO:-LPCvLeYCk_j94ySL1Cr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/df4c452d652d271a0bc927cd184802c50bca7b23/pkg/sql/opt/exec/execbuilder/relational_builder.go#L291)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/IF/If/
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> for the SQL changes.
The overall approach makes sense to me with my limited understanding of the KV layer, but I can't be the only person to sign off on this PR.

I do recommend adding a comment to explain why the limit participates in the test, to highlight that it's not strictly required for correctness.

> Reviewed 1 of 3 files at r1, 17 of 17 files at r5, 17 of 17 files at r6.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LPRA0rHAc-3lEOlo06w:bf3d03b)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/scan.go, line 264 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPR8rqu4Fa2O9V0J3df:-LPR8rqu4Fa2O9V0J3dg:bfm7hbj) ([raw file](https://github.com/cockroachdb/cockroach/blob/a7ddb4495ddd6d86918041da6c077af6bef8cdac/pkg/sql/scan.go#L264)):*
> ```Go
> // distSender level safely.
> func (n *scanNode) canParallelize() bool {
> 	return n.allPointSpans && n.limitHint() == 0
> ```

add a comment here to remind the reader why limitHint == 0 participates in this test and why it was added.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
I would also recommend a cluster setting to opt-out of this optimization if it's intended for use with 2.1.
Since we haven't had much time to test this in prod an opt-out will offer a road out for users encountering perf problems.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LPRAfo80jQIEtsNmYpn:bly34n4)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

RaduBerinde commented on pkg/sql/distsqlrun/tablereader.go:
```diff
@@ -245,9 +249,18 @@ func (tr *tableReader) Start(ctx context.Context) context.Context {
+	// tableReader spans are point spans. This enables distsender parallelism -
+	// if limitBatches is true, distsender does *not* parallelize multi-range
+	// scan requests.
+	if tr.allPointSpans && tr.limitHint == 0 {
```
Maybe only do this if we don't have a ton of spans?

jordanlewis replied:
Seems like that would be taking us even deeper into random heuristic territory. What would you propose as a cutoff? I think we should just hand it off to distSender, which will naturally limit parallelism for us.

RaduBerinde replied:
Not really worried about parallelism, but about having to hold all results in memory at once.. I'd use the default batch size as a cutoff.

jordanlewis replied:
But then beyond that cutoff, you'd have a horrible performance degradation: you'd revert to fully serialized execution. Well, I'll add it as protection I suppose.

jordanlewis replied:
Actually I still think this doesn't make too much sense. What "default batch size" are you talking about, btw?

RaduBerinde replied:
`kvBatchSize` in kvfetcher.go. I don't feel strongly about this (you can make the argument that the size of the KVs is proportional to the size of the spans which must be in memory regardless), your call.

RaduBerinde commented:
Cluster setting is a great idea, just in case.

bdarnell commented:
Do we ever coalesce adjacent spans? That is, would we turn `SELECT * FROM t WHERE x in (1, 2)` from two point lookups into a range span? That could break this optimization.

RaduBerinde commented:
It does not consider `[/1 - /1]` and `[/2 - /2]` adjacent. In any case, at worst that would disable the optimization - the `parallel` flag is only set if all the spans have start key = end key.

bdarnell commented:
Whether we merge adjacent spans or not today, we might in the future, so we should at least add a test for that case so we won't break it without realizing.

RaduBerinde commented:
If you backport this, please backport #31662 as well (or ping me and I'll do it).

jordanlewis commented:
Hmm, looks like we *do* coalesce adjacent spans.

```
                        testdata/planner_test/select:417: EXPLAIN SELECT * FROM a WHERE a = 1 OR a = 2
                        expected:
                            scan  ·         ·
                            ·     table     a@primary
                            ·     spans     /1-/1/# /2-/2/#
                            ·     parallel  ·

                        but found (query options: "") :
                            scan  ·      ·
                            ·     table  a@primary
                            ·     spans  /1-/2/#
```

jordanlewis commented:
Added a cluster setting, `sql.parallel_point_scans.enabled`.

RaduBerinde commented:
Oh.. I looked at the code, indeed we added a ConsolidateSpans post-processing step when generating spans; I had forgotten about that (sorry).

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1077 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LPW4r4u9tyiioULcx6A:-LPW4r4u9tyiioULcx6B:b6rcus6) ([raw file](https://github.com/cockroachdb/cockroach/blob/19da10a267727248553baf00694e3f4f4cc3a26d/pkg/sql/opt/exec/execbuilder/testdata/select#L1077)):*
> ```
> 
> # This one isn't parallelizable because its index has nullable values.
> # TODO(radu): this shouldn't be parallelizable: c has nullable values.
> ```

It doesn't, because it's either 10 or 20. If you had `c IS NULL OR c = 20` it shouldn't be parallelizable.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
> It doesn't, because it's either 10 or 20. If you had c IS NULL OR c = 20 it shouldn't be parallelizable.

Oh, that makes a lot of sense. I didn't realize that the optimizer was examining the actual constraint values to determine nullability. I'll fix the test.

asubiotto commented:
Extended this PR to work in the case where all the columns are specified in the span keys apart from the last one. In that case, only if we're dealing with ints, we calculate a difference to get the maximum possible size of the result set. Note that doing this for all other types is kind of difficult but since it's difficult, point spans won't be consolidated (as it's very unlikely, for example, to have a float point span where the `.Next()` float is another point span). The only queries that won't benefit from parallelization are range queries on non-int types (e.g. `a > x and a < y`).

PTAL.

asubiotto commented:
ping

jordanlewis commented:
Taking a look today. I'd like another review from @nvanbenschoten if he doesn't mind as well.

nvanbenschoten commented:
Yeah I'll take a look. Thanks for working on this @asubiotto!

asubiotto commented:
Thanks! Fixed up the lint and logictest issues (just needed to be updated)

asubiotto commented:
```
root@localhost:26257/kv> SELECT v FROM kv WHERE k IN (-2967608882202863871, 9141638872723910047);
  v
+---+
  f
  (
(2 rows)

Time: 483.627577ms
root@localhost:26257/kv> SET CLUSTER SETTING sql.parallel_scans.enabled=true;
SET CLUSTER SETTING

Time: 741.784097ms

root@localhost:26257/kv> SELECT v FROM kv WHERE k IN (-2967608882202863871, 9141638872723910047);
  v
+---+
  f
  (
(2 rows)

Time: 242.030463ms
```

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt_index_selection.go, line 317 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSriU1gEPqcCEidEyoR:-LSriU1gEPqcCEidEyoS:b-42i8p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt_index_selection.go#L317)):*
> ```Go
> 	}
> 	// TODO(asubiotto): I don't get this condition, couldn't the columns to read
> 	// be a subset of the index?
> ```

If the columns that are constrained are only a subset of the index, then there's no guarantee on how many rows we return - since the unconstrained part of the index could have an unbounded cardinality.
___
*[pkg/sql/scan.go, line 272 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSriqpE9PUsrDkhvQCK:-LSriqpF62iGqLC9OFXv:b3kre) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L272)):*
> ```Go
> 	// We can't parallelize if we have a non-zero limit hint, since distsender
> 	// is limited to running limited batches serially.
> 	// TODO(asubiotto): But this limit refers to the output limit?
> ```

?
___
*[pkg/sql/distsqlrun/tablereader.go, line 265 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrixaD3pegZvHd1Fy8:-LSrixaD3pegZvHd1Fy9:b-o5i0gg) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/distsqlrun/tablereader.go#L265)):*
> ```Go
> 	// This enables distsender parallelism - if limitBatches is true, distsender
> 	// does *not* parallelize multi-range scan requests.
> 	// TODO(asubiotto): I'm not sure this limitHint == 0 check is necessary.
> ```

It's necessary if we want to preserve the "cautious" behavior of the distsender - to never parallelize multiple requests when the top level request has a limit, in case a single parallel request satsifies the top level limit all on its own.
___
*[pkg/sql/opt/constraint/constraint.go, line 581 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrjip112WBt4Q2LbFo:-LSrjip112WBt4Q2LbFp:b-xna8fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/constraint/constraint.go#L581)):*
> ```Go
> 	// same start and end values covers all columns.
> 	prefix := c.Prefix(evalCtx)
> 	distinctVals := uint64(1)
> ```

This seems wrong - we're assuming the # of distinct values is 1, but then below we don't update it for types of the last column != int. Am I missing something?
___
*[pkg/sql/opt/constraint/constraint.go, line 586 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrjxEJDwLGDY4ZG-8I:-LSrjxEJDwLGDY4ZG-8J:byqtvub) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/constraint/constraint.go#L586)):*
> ```Go
> 	} else if prefix == numCols-1 {
> 		// If the prefix does not include the last column, calculate the number of
> 		// distinct values possible in the span.
> ```

In other words, if we don't have the last value == int, don't we have to return 0 here?
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1082 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrk9vC-6Qo9M58TVSj:-LSrk9vC-6Qo9M58TVSk:bk9m7sc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/exec/execbuilder/testdata/select#L1082)):*
> ```
> # ------------------------------------------------------
> # Verify that multi-span point lookups are parallelized.
> # ------------------------------------------------------
> ```

Can you add tests for non-int datatypes?


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> Reviewed 1 of 21 files at r9, 3 of 6 files at r10, 27 of 27 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)



<!-- Sent from Reviewable.io -->

nvb reviewed (changes requested):
> Reviewed 2 of 17 files at r5, 15 of 21 files at r9, 27 of 27 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/scan.go, line 118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuS6-Y47xM217yA1R2:-LSuS6-Y47xM217yA1R3:bt4c23i) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L118)):*
> ```Go
> 
> 	// Set if the scanNode's spans are each guaranteed to return at most one row.
> 	maxResultHint uint64
> ```

Explain what this is, not just when it is set.
___
*[pkg/sql/scan.go, line 270 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuRFdL0kB-39X2DDds:-LSuRFdL0kB-39X2DDdt:byquctd) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L270)):*
> ```Go
> 	// ParallelScanResultThreshold results will be returned, to prevent potential
> 	// memory blowup.
> 	// We can't parallelize if we have a non-zero limit hint, since distsender
> ```

s/distsender/DistSender/
___
*[pkg/sql/logictest/testdata/planner_test/select, line 492 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuSiL675bSlkZ1y71R:-LSuSiL675bSlkZ1y71S:b-q7ufpy) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/logictest/testdata/planner_test/select#L492)):*
> ```
> CREATE UNIQUE INDEX on b(c)
> 
> # This one isn't parallelizable because its index has nullable values.
> ```

This is an interesting condition. Why can't we parallelize when an index has nullable values? We can still derive an upper-bound on the number of rows returned, right?
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1082 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrk9vC-6Qo9M58TVSj:-LSuT8BQ6qKHNtAzJPEX:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/exec/execbuilder/testdata/select#L1082)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Can you add tests for non-int datatypes?
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt_index_selection.go, line 317 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSriU1gEPqcCEidEyoR:-LSvCVqA83ARo_dmPzCO:b-zcjvte) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt_index_selection.go#L317)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

If the columns that are constrained are only a subset of the index, then there's no guarantee on how many rows we return - since the unconstrained part of the index could have an unbounded cardinality.
</blockquote></details>

Thanks
___
*[pkg/sql/scan.go, line 118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuS6-Y47xM217yA1R2:-LSvDX8ZA0sscSbDzVD0:bv07bvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L118)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Explain what this is, not just when it is set.
</blockquote></details>

Done. This comment was outdated and I think it's clearer to just explain what it is.
___
*[pkg/sql/scan.go, line 270 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuRFdL0kB-39X2DDds:-LSvDp-_7ehTijpK38jL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L270)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

s/distsender/DistSender/
</blockquote></details>

Done.
___
*[pkg/sql/scan.go, line 272 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSriqpE9PUsrDkhvQCK:-LSvDuxS7CtnaUFESgfD:b-aldee1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L272)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

?
</blockquote></details>

Same comment as below (about limitHint == 0).
___
*[pkg/sql/distsqlrun/tablereader.go, line 265 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrixaD3pegZvHd1Fy8:-LSvEAlVCVb7FJmZBn_-:b-3bey4h) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/distsqlrun/tablereader.go#L265)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

It's necessary if we want to preserve the "cautious" behavior of the distsender - to never parallelize multiple requests when the top level request has a limit, in case a single parallel request satsifies the top level limit all on its own.
</blockquote></details>

But it seems that prior to this, cautious behavior meant not blowing up memory usage. While now, given that we know the result size, the worst case would be overfetching. I could be missing something though.
___
*[pkg/sql/logictest/testdata/planner_test/select, line 492 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuSiL675bSlkZ1y71R:-LSyvJyH5jswqKBm7kJ6:bea59i6) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/logictest/testdata/planner_test/select#L492)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

This is an interesting condition. Why can't we parallelize when an index has nullable values? We can still derive an upper-bound on the number of rows returned, right?
</blockquote></details>

That's an interesting point. I guess we can be more fine-grained and not parallelize only when we are not reading null values from a nullable column (e.g. this case). However, we can't do so when we read nulls since you can have multiple nulls even on a unique index i.e.:
```
root@:26257/defaultdb> CREATE TABLE b (a INT, c INT NULL, UNIQUE INDEX a (a, c));
CREATE TABLE

Time: 7.653ms

root@:26257/defaultdb> INSERT INTO b VALUES (1, NULL), (1, NULL);
INSERT 2

Time: 6.958ms
```
I'll look into parallelizing when no nulls are specified when reading from a nullable column.
___
*[pkg/sql/opt/constraint/constraint.go, line 581 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrjip112WBt4Q2LbFo:-LSv85eX-L6uYaJLEoR0:b-3l5uct) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/constraint/constraint.go#L581)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This seems wrong - we're assuming the # of distinct values is 1, but then below we don't update it for types of the last column != int. Am I missing something?
</blockquote></details>

No, thanks for catching that. Added a test case that exercises this logic.
___
*[pkg/sql/opt/constraint/constraint.go, line 586 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrjxEJDwLGDY4ZG-8I:-LSv8ARPCfN9luZnGLj4:bgwt35a) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/constraint/constraint.go#L586)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

In other words, if we don't have the last value == int, don't we have to return 0 here?
</blockquote></details>

Yes. Fixed.
___
*[pkg/sql/opt/exec/execbuilder/testdata/select, line 1082 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrk9vC-6Qo9M58TVSj:-LSvB5BkAJKbrgSNAGu5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/opt/exec/execbuilder/testdata/select#L1082)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto commented:
This is ready for another look. @nvanbenschoten, added logic for NULLs.

jordanlewis reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LT-mmq403Ncp5ZQILnu:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt_index_selection.go, line 342 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LT-kEU61oCF6UwbUnre:-LT-kEU61oCF6UwbUnrf:b-16v6vq) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/opt_index_selection.go#L342)):*
> ```Go
> 		// unconstrained.
> 		for i := 0; i < c.Spans.Count(); i++ {
> 			start := c.Spans.Get(i).StartKey()
> ```

Is it sufficient to just check the start key? is there anything weird with descending indexes? I feel like if the index is descending the span might be in reverse.

I played around with this and couldn't find a way to get NULL in the EndKey and anything but in the StartKey, so I think we're fine. But just wanted to put it out there.
___
*[pkg/sql/distsqlrun/tablereader.go, line 265 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrixaD3pegZvHd1Fy8:-LT-jIHw9G012BL0Lzs7:bywatzr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/distsqlrun/tablereader.go#L265)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

But it seems that prior to this, cautious behavior meant not blowing up memory usage. While now, given that we know the result size, the worst case would be overfetching. I could be missing something though.
</blockquote></details>

That's a good point - I think it makes sense to remove that condition.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt_index_selection.go, line 342 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LT-kEU61oCF6UwbUnre:-LT3Ak3xCanD0Gm4q8Bi:bs1zfcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/opt_index_selection.go#L342)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Is it sufficient to just check the start key? is there anything weird with descending indexes? I feel like if the index is descending the span might be in reverse.

I played around with this and couldn't find a way to get NULL in the EndKey and anything but in the StartKey, so I think we're fine. But just wanted to put it out there.
</blockquote></details>

@RaduBerinde what do you think?


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/distsqlrun/tablereader.go, line 265 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSrixaD3pegZvHd1Fy8:-LT3UrSZCM5M3SifMR8R:byg97el) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/distsqlrun/tablereader.go#L265)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

That's a good point - I think it makes sense to remove that condition.
</blockquote></details>

I'll do so in a separate PR.


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> from my perspective, but I didn't dig into the optimizer stuff. I'm really impressed by all of this.

> Reviewed 5 of 27 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LT4LJDs5Bw7XvSw2SbD:b-2r7zfm)* status: :shipit: complete! 2 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/scan.go, line 118 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuS6-Y47xM217yA1R2:-LT4KlF82ONNGviXPihe:b6lzyra) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/scan.go#L118)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Done. This comment was outdated and I think it's clearer to just explain what it is.
</blockquote></details>

Thanks, that's much better.
___
*[pkg/sql/logictest/testdata/planner_test/select, line 492 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LSuSiL675bSlkZ1y71R:-LT4L16u3Yqlm4dURhzz:bq1loks) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc7069a291f343081ed6c83bbc4ac9b8b78fc0e7/pkg/sql/logictest/testdata/planner_test/select#L492)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

That's an interesting point. I guess we can be more fine-grained and not parallelize only when we are not reading null values from a nullable column (e.g. this case). However, we can't do so when we read nulls since you can have multiple nulls even on a unique index i.e.:
```
root@:26257/defaultdb> CREATE TABLE b (a INT, c INT NULL, UNIQUE INDEX a (a, c));
CREATE TABLE

Time: 7.653ms

root@:26257/defaultdb> INSERT INTO b VALUES (1, NULL), (1, NULL);
INSERT 2

Time: 6.958ms
```
I'll look into parallelizing when no nulls are specified when reading from a nullable column.
</blockquote></details>

Perfect, that avoids what would have been a really edge corner case to explain.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> modulo the descending columns issue

> Reviewed 4 of 23 files at r3, 1 of 20 files at r8, 4 of 31 files at r12, 27 of 27 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LTNkMYl1JtQnOoznDmX:b-flkfe0)* status: :shipit: complete! 3 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt_index_selection.go, line 342 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LT-kEU61oCF6UwbUnre:-LTNiiPT6loFdSOL21H-:bv2wfn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/opt_index_selection.go#L342)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

@RaduBerinde what do you think?
</blockquote></details>

We'd need to look at the end key for columns that are descending.. E.g. https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/idxconstraint/testdata/single-column#L100

I'd make sure to add tests like this (including cases where we have multiple columns with varying directions; and for both implementations of this logic).

Note that there is a function `Constraint.ExtractNotNullCols` which gives you all not-null columns implied by the constraint, I think you can just use that (in conjunction with the info on what columns are nullable in the schema).
___
*[pkg/sql/opt_index_selection.go, line 344 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNjaq50C8cHqoSg-r3:-LTNjaq50C8cHqoSg-r4:b-p6lgwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/opt_index_selection.go#L344)):*
> ```Go
> 			start := c.Spans.Get(i).StartKey()
> 			for j := 0; j < start.Length(); j++ {
> 				if start.Value(j) == tree.DNull {
> ```

This could be an exclusive key (e.g. `(NULL - ]`) which would be fine (need to check `StartBoundary()`
___
*[pkg/sql/scan.go, line 102 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNiCLIDD9WjWkHdqhv:-LTNiCLIDD9WjWkHdqhw:bweom7v) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/scan.go#L102)):*
> ```Go
> 	disableBatchLimits bool
> 
> 	// Should be set to true if distsqlrun.ParallelScans is true.
> ```

`sqlbase.ParallelScans`
___
*[pkg/sql/distsqlrun/processors.proto, line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNhGMH-vzDpkseQ6Y4:-LTNhGMH-vzDpkseQ6Y5:b-gd52vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/distsqlrun/processors.proto#L231)):*
> ```protobuf
>   optional ScanVisibility visibility = 7 [(gogoproto.nullable) = false];
> 
>   // If non-zero, this is a hint for the maximum number of rows a TableReader
> ```

[nit] here also, provide more information (explain that it's a guaranteed upper bound).
___
*[pkg/sql/opt/constraint/constraint.go, line 569 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNfcobE77Xrp7LxaAb:-LTNfcoc5BiMwpsKNgfl:bwmm01t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/constraint/constraint.go#L569)):*
> ```Go
> }
> 
> // CalculateMaxResults returns a non-zero integer indicating the maximum number
> ```

this function has plenty of `return 0`s, need to explain the zero value here
___
*[pkg/sql/opt/exec/factory.go, line 57 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNgr2KAwgVCDT6dON8:-LTNgr2KAwgVCDT6dON9:bsekc8g) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/exec/factory.go#L57)):*
> ```Go
> 	//     the scan.
> 	//
> 	// maxResultHint may be 0 if the maximum number of results is unknown.
> ```

[nit] I'd add more explanation here. In particular "hint" usually suggests there are no guarantees, but this is guaranteed to be an upper bound.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 277 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNg7rr1ii3Jgjj4yXQ:-LTNg7rr1ii3Jgjj4yXR:bh1xwm5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/exec/execbuilder/relational_builder.go#L277)):*
> ```Go
> 
> // indexConstraintMaxResults returns the maximum number of results that a scan
> // is guaranteed to read. Iff this hint is invalid, 0 is returned.
> ```

nit: "is guaranteed to read" is strange phrasing. Maybe "maximum possible number of results for a scan; the scan is guaranteed never to return more results than this".
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 305 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNgQ0b9ch6ucm8UkJm:-LTNgQ0b9ch6ucm8UkJn:btmzuov) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/exec/execbuilder/relational_builder.go#L305)):*
> ```Go
> 		if !indexCols.SubsetOf(rel.NotNullCols) {
> 			// Ensure that if we have nullable columns, we are only reading non-null
> 			// values, given that the number of nulls in a unique index is
> ```

[nit]: more accurate phrasing would be "given that a unique index allows arbitrary number of duplicate entries if they have NULLs"


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
Fixed descending columns issue and added tests. Merging this soon but holding off to allow for any final comments. Thanks for all the reviews!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-:-LTSWJoW36B16r2IdFZg:b-msyjzl)* status: :shipit: complete! 3 of 0 LGTMs obtained (and 1 stale)

___
*[pkg/sql/opt_index_selection.go, line 342 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LT-kEU61oCF6UwbUnre:-LTS3XTv-GomkgXzqoTv:b-57796q) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/opt_index_selection.go#L342)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

We'd need to look at the end key for columns that are descending.. E.g. https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/idxconstraint/testdata/single-column#L100

I'd make sure to add tests like this (including cases where we have multiple columns with varying directions; and for both implementations of this logic).

Note that there is a function `Constraint.ExtractNotNullCols` which gives you all not-null columns implied by the constraint, I think you can just use that (in conjunction with the info on what columns are nullable in the schema).
</blockquote></details>

Done. Thanks for the pointer, using `ExtractNotNullCols` and checking if the index cols are a subset of those.
___
*[pkg/sql/opt_index_selection.go, line 344 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNjaq50C8cHqoSg-r3:-LTS4rmHBRqn-7JpobJ0:b-73as3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/opt_index_selection.go#L344)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This could be an exclusive key (e.g. `(NULL - ]`) which would be fine (need to check `StartBoundary()`
</blockquote></details>

Using `ExtractNotNullCols` which does so (and tested this). Returning 0 in this case
___
*[pkg/sql/scan.go, line 102 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNiCLIDD9WjWkHdqhv:-LTSGMNy-r86cOSXQxcN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/scan.go#L102)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

`sqlbase.ParallelScans`
</blockquote></details>

Done.
___
*[pkg/sql/distsqlrun/processors.proto, line 231 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNhGMH-vzDpkseQ6Y4:-LTSGx5WDLC4I4yEoLv2:bs7wumx) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d624402bfa69025594b56442d219f37a06854f6/pkg/sql/distsqlrun/processors.proto#L231)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] here also, provide more information (explain that it's a guaranteed upper bound).
</blockquote></details>

Done. 
___
*[pkg/sql/opt/constraint/constraint.go, line 569 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNfcobE77Xrp7LxaAb:-LTSHAc38wT2F8y25tcD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/constraint/constraint.go#L569)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

this function has plenty of `return 0`s, need to explain the zero value here
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/factory.go, line 57 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNgr2KAwgVCDT6dON8:-LTSJe566cf0lyk5RYfY:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/exec/factory.go#L57)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I'd add more explanation here. In particular "hint" usually suggests there are no guarantees, but this is guaranteed to be an upper bound.
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 277 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNg7rr1ii3Jgjj4yXQ:-LTSK-oh4RMM6rVOpJXt:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/exec/execbuilder/relational_builder.go#L277)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

nit: "is guaranteed to read" is strange phrasing. Maybe "maximum possible number of results for a scan; the scan is guaranteed never to return more results than this".
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/relational_builder.go, line 305 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTNgQ0b9ch6ucm8UkJm:-LTSKBRk-Db6Gr_D7ssH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e85d32e9bfb3bec6baacbe49096c3eb645162693/pkg/sql/opt/exec/execbuilder/relational_builder.go#L305)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit]: more accurate phrasing would be "given that a unique index allows arbitrary number of duplicate entries if they have NULLs"
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 4 stale)

___
*[pkg/sql/opt/constraint/constraint.go, line 589 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTT8Hga9RIBVwc8ABIh:-LTT8Hga9RIBVwc8ABIi:bdw7syy) ([raw file](https://github.com/cockroachdb/cockroach/blob/14c75cf9288a5354eda159c9903881c907b7e269/pkg/sql/opt/constraint/constraint.go#L589)):*
> ```Go
> 	// values, given that a unique index allows an arbitrary number of duplicate
> 	// entries if they have NULLs.
> 	if hasNullableCols && !indexCols.SubsetOf(c.ExtractNotNullCols(evalCtx)) {
> ```

This is perhaps a bit pedantic, but it's possible that some of the cols are not-nullable, and the others are not null in the constraint. I would pass the list of known not null cols instead of a single flag, and union that with the extracted not null cols.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31616)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 4 stale)

___
*[pkg/sql/opt/constraint/constraint.go, line 589 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/31616#-LTT8Hga9RIBVwc8ABIh:-LTTy4_qE7nHlpqLmSb3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/14c75cf9288a5354eda159c9903881c907b7e269/pkg/sql/opt/constraint/constraint.go#L589)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is perhaps a bit pedantic, but it's possible that some of the cols are not-nullable, and the others are not null in the constraint. I would pass the list of known not null cols instead of a single flag, and union that with the extracted not null cols.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

asubiotto commented:
Just to highlight the improvement. I ran `./workload run kv --batch=100 --splits=1000 --read-percent=95 --duration=1m` on a 3-node roachprod cluster with `120ms` of artificial network delay.

without parallel scans enabled (after truncating kv)

```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   60.0s        0             72            1.2   6307.1   6442.5   7516.2   7784.6   7784.6  read
   60.0s        0              5            0.1    755.0    771.8    906.0    906.0    906.0  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
   60.0s        0             77            1.3   5946.5   6442.5   7516.2   7784.6   7784.6
```

with parallel scans enabled:
```
_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
   60.0s        0            873           14.5    507.7    385.9   1006.6   1208.0   1744.8  read
   60.0s        0             38            0.6    835.3    872.4    906.0   1006.6   1006.6  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
   60.0s        0            911           15.2    521.4    385.9   1006.6   1208.0   1744.8
```

asubiotto commented:
bors r+

jordanlewis commented:
Very nice.
